bundles/org.openhab.binding.enigma2/src/main/java/org/openhab/binding/enigma2/internal/Enigma2Client.java

```diff
@@ -0,0 +1,336 @@
+    static final String PATH_EPG = "/web/epgservicenow?sRef=";
+    static final String PATH_ANSWER = "/web/messageanswer?getanswer=now";
+    static final int TYPE_QUESTION = 0;
+    static final int TYPE_INFO = 1;
+    static final int TYPE_WARNING  = 2;
+    static final int TYPE_ERROR = 3;
+    private Map<String,@Nullable String> channels = new LinkedHashMap<>();
+    private String host;
+    private boolean power;
+    private String channel = "";
+    private String title = "";
+    private String description = "";
+    private String answer = "";
+    private int volume = 0;
+    private int timeout;
+    private boolean mute;
+    private boolean online;
+    private boolean initialized = false;
+    private boolean asking = false;
+    private LocalDateTime lastAnswerTime = LocalDateTime.of(2020,1,1, 0,0); // Date in the past
+
+    public Enigma2Client(String host, @Nullable String user, @Nullable String password, int timeout) {
+        this.timeout = timeout;
+        if(StringUtils.isNotEmpty(user) && StringUtils.isNotEmpty(password)) {
+            this.host = "http://" + user + ":" + password + "@" + host;
+        } else {
+            this.host = "http://" + host;
+        }
+    }
+
+    public boolean refresh() {
+        boolean wasOnline = online;
+        refreshPower();
+        if(!wasOnline && online) {
+            // Only refresh all services if the box changed from offline to online and power is on
+            // because it is a performance intensive action.
+            refreshAllServices();
+        }
+        refreshChannel();
+        refreshEpg();
+        refreshVolume();
+        refreshAnswer();
+        return online;
+    }
+
+    public void refreshPower() {
+        Optional<Document> document = transmit(PATH_POWER);
+        if(document.isPresent()) {
+            online = true;
+            processPowerResult(document.get());
+        } else {
+            online = false;
+            power = false;
+        }
+        initialized = true;
+    }
+
+    public void refreshAllServices() {
+        if(power || channels.isEmpty()) {
+            transmit(PATH_ALL_SERVICES).ifPresent(this::processAllServicesResult);
+        }
+    }
+
+    public void refreshChannel() {
+        if(power) {
+            transmit(PATH_CHANNEL).ifPresent(this::processChannelResult);
+        }
+    }
+
+    public void refreshAnswer() {
+        if(asking) {
+            transmit(PATH_ANSWER).ifPresent(this::processAnswerResult);
+        }
+    }
+
+    public void refreshVolume() {
+        if(power) {
+            transmit(PATH_VOLUME).ifPresent(this::processVolumeResult);
+        }
+    }
+
+    public void refreshEpg() {
+        if(power) {
+            Optional.ofNullable(channels.get(channel)).flatMap(name -> transmit(PATH_EPG + UrlEncoded.encodeString(name))).ifPresent(this::processEpgResult);
+        }
+    }
+
+    private Optional<Document> transmit(String path) {
```

Comment:
Since not all transmit calls even care about the return result, and parsing xml is very computationally heavy, I think it would be better to allow the caller to pass in a Consumer instead. That way the `transmit` method can skip parsing the result for callers that don't care about it.

```suggestion
    private Optional<Document> transmit(String path, @Nullable Consumer<Document> consumer) {
```

Reply:
Fixed, but not with a Consumer. Instead I divided it into two methods.